Sources/FoundationEssentials/AttributedString/AttributedString+UTF8View.swift

```diff
@@ -0,0 +1,139 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if FOUNDATION_FRAMEWORK
+@_spi(Unstable) internal import CollectionsInternal
+#elseif canImport(_RopeModule)
+internal import _RopeModule
+#elseif canImport(_FoundationCollections)
+internal import _FoundationCollections
+#endif
+
+@available(FoundationPreview 6.2, *)
+extension AttributedString {
+    public struct UTF8View: Sendable {
+        internal var _guts: Guts
+        internal var _range: Range<BigString.Index>
+        internal var _identity: Int = 0
+
+        internal init(_ guts: AttributedString.Guts) {
+            self.init(guts, in: guts.stringBounds)
+        }
+
+        internal init(_ guts: Guts, in range: Range<BigString.Index>) {
+            _guts = guts
+            _range = range
+        }
+
+        public init() {
+            self.init(Guts())
+        }
+    }
+
+    public var utf8: UTF8View {
+        UTF8View(_guts)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension AttributedSubstring {
+    public var utf8: AttributedString.UTF8View {
+        AttributedString.UTF8View(_guts, in: _range)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension AttributedString.UTF8View {
+    var _utf8: BigSubstring.UTF8View {
+        BigSubstring.UTF8View(_unchecked: _guts.string, in: _range)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension AttributedString.UTF8View: BidirectionalCollection {
+    public typealias Element = UTF8.CodeUnit
+    public typealias Index = AttributedString.Index
+    public typealias Subsequence = Self
+
+    public var startIndex: AttributedString.Index {
+        .init(_range.lowerBound)
+    }
+
+    public var endIndex: AttributedString.Index {
+        .init(_range.upperBound)
+    }
+
+    public var count: Int {
+        _utf8.count
+    }
+
+    public func index(before i: AttributedString.Index) -> AttributedString.Index {
+        precondition(i >= startIndex && i <= endIndex, "AttributedString index out of bounds")
+        let j = Index(_guts.string.utf8.index(before: i._value))
```

Comment:
Does this throw if `i == startIndex`? I believe calling `string.utf8.index(before: string.utf8.startIndex)` throws the index out of bound error. If so, should we just enforce this at the precondition above?

Reply:
We still do hit a precondition from swift-collections:

```
_RopeModule/BigString+Contents.swift:292: Precondition failed: Can't advance below start index
```

But we might as well account for this here, good call - I'll update the above `>=` to be `>`